lib/acts_as_taggable_on_mongoid/taggable/tag_type_definition.rb

```diff
@@ -0,0 +1,188 @@
+
+        owner.taggable_mixin.module_eval do
+          break if methods.include?(base_tags_method)
+
+          define_method base_tags_method do
+            tag_definition.tags_table.where(taggable_type: tag_definition.owner.name)
+          end
+        end
+      end
+
+      def define_relations
+        # Relations cannot be added for the tags and taggings like they are in ActiveRecord because
+        # Mongoid does not allow for a scope like ActiveRecord does.
+        #
+        # Therefore the relation like actions will have to be defined separately ourselves.  If any
+        # relation actions are missed, we'll just have to fix it here when we find them.
+        # (This is far from ideal, but it is the only way to work around the issue at this time.)
+
+        add_context_taggings_method
+        add_context_tags_method
+      end
+
+      def add_context_taggings_method
+        tag_definition = self
+
+        owner.taggable_mixin.module_eval do
+          define_method "#{tag_definition.single_tag_type}_taggings".to_sym do
+            send(tag_definition.taggings_name).
+                order_by(*tag_definition.taggings_order).
+                for_tag(tag_definition)
+          end
+        end
+      end
+
+      def add_context_tags_method
+        tag_definition = self
+
+        owner.taggable_mixin.module_eval do
+          define_method tag_definition.tag_type.to_sym do
+            send(tag_definition.base_tags_method).
+                order_by(*tag_definition.taggings_order).
+                for_tag(tag_definition)
+          end
+        end
+      end
+
+      def add_tag_list
+        add_list_getter
+        add_list_setter
+        add_all_list_getter
+        add_list_exists
+        add_list_change
+        add_list_changed
+        add_changed_from_default?
+        add_will_change
+        add_get_was
+        add_reset_list
+      end
+
+      def add_list_getter
+        tag_definition = self
+        tag_list_name  = tag_definition.tag_list_name
+
+        owner.taggable_mixin.module_eval do
+          define_method(tag_list_name) do
+            tag_list_on tag_definition
+          end
+
+          alias_method "#{tag_list_name}_before_type_cast".to_sym, tag_list_name.to_sym
+        end
+      end
+
+      # TODO: Refactor this so that dirty attributes are set properly.
+      def add_list_setter
+        tag_definition = self
+
+        owner.taggable_mixin.module_eval do
+          define_method("#{tag_definition.tag_list_name}=") do |new_tags|
+            new_tags = Array.wrap(new_tags)
+            options  = new_tags.extract_options!
+
+            options.merge!(parse: true)
+
+            new_list = tag_definition.parse(*new_tags, options)
+
+            mark_tag_list_changed(new_list)
+            tag_list_set(new_list)
+          end
+        end
+      end
+
+      def add_all_list_getter
+        tag_definition = self
+
+        owner.taggable_mixin.module_eval do
```

Comment:
OK, I admit - adding all of these methods dynamically with define_method is a bit ugly.

But it is WAY less ugly than adding them as a giant string blob like ActsAsTaggableOn does.

For what little it is worth, this exact method is what Mongoid does/uses to define field methods like `field_name?`, `field_name_was`, etc. so it may be a bit ugly, but it is a common practice, so I'm keeping it.